### ISSUE TITLE
Image Preprocessing takes a lot of time

### ISSUE THREAD

**Issue by smessica** [jrzaurin/pytorch-widedeep#168]
Hello, Javier!
First I wanted to say I really love pytorch-widedeep and think it's great!

I tried to use it to create and train a multimodal model that takes images and tabular data and returns predictions based on both sources. 
In order to do that, I used the image preprocessor on 5000 images and it takes a lot of time. is it normal? or do I have some problem? 
I think that the reading itself in the ImagePreprocessor under the transform function is causing the time problem:
imgs = [cv2.imread("/".join([self.img_path, img])) for img in image_list]

Do you support using custom pre-processors?

Thanks,
Shvat

**Comment by jrzaurin** [comment]
Hi @smessica 

thanks for your comment! I wish I had more time these days to add more features other than maintaining it. 

Regarding to `ImagePreprocessor`, how big are the images? reading should not be that much of a problem.

Nonetheless, the library does support your custom preprocessor, yes 🙂. 

If you have a look here we have a (very) simple `[BasePreprocessor](https://github.com/jrzaurin/pytorch-widedeep/blob/05e007e9913f22e34c4606b14e6f6bb30ca2fbc1/pytorch_widedeep/preprocessing/base_preprocessor.py#L10)`  that you can use and then do whatever operation is it that you want within the corresponding methods. 

One thing that I want to do is improving typing. As you will see, the BasePreprocessor does not provide info regarding the format of the output for each of its methods (`fit`, `transform` and `fit_transform`) but the output must be an array.

So in summary (as a form of pseudocode): 

```python
from pytorch_widedeep.preprocessing.base_preprocessor import BasePreprocessor

class MyPreProcessor(BasePreprocessor):
    def __init__(self, ...):
        init_stuff

    def fit(self, df: pd.DataFrame) -> np.ndarray:
        do fit_stuff and return array      

    def transform(self, df: pd.DataFrame) -> np.ndarray:
        do transform_stuff and return array    

    def fit_transform(self, df: pd.DataFrame) -> np.ndarray:
        do fit_and_transform_stuff and return array   
```

Let me know if this helps

**Comment by smessica** [comment]
Thank you for the quick response! (:
The images are retina images and are 4000x4000. 
I end up doing the solution you suggested in the issue: "Working with large datasets" for memory-efficient reasons.
It works great!

Thanks again,
Shvat
